mds: fix stray evaluation using scrub and introduce new option

dparmar18 opened the pull request:
This PR fixes the OOM failure caused by race between Purge and Scrub while also
introduces a new option `scrub_mdsdir` to evaluate strays at root.

Fixes: https://tracker.ceph.com/issues/53724, https://tracker.ceph.com/issues/51824
Signed-off-by: Dhairya Parmar <dparmar@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

mchangir commented:
@dparmar18 some description about what the problem is and why this solution of evaluating strays is beneficial would be prudent

vshankar commented on src/mds/ScrubStack.cc:
```diff
@@ -117,6 +117,7 @@ int ScrubStack::enqueue(CInode *in, ScrubHeaderRef& header, bool top)
   clog_scrub_summary(in);
 
   kick_off_scrubs();
+  in->mdcache->maybe_eval_stray(in, false);
```
This needs to be done as part of scrubbing the inode and not during enqueue.

lxbsz replied:
IMO you should iterate and check all the dentries in the stray directories. Just call this for that ?

dparmar18 replied:
No need to iterate as scrub already could do it, it's just that it was bugged and would go into a race condition which has has been fixed in this PR.

vshankar commented:
@dparmar18 This doesn't seem to address the issue mentioned in the tracker. What prompted you to introduce this change?

dparmar18 commented:
> @dparmar18 This doesn't seem to address the issue mentioned in the tracker. What prompted you to introduce this change?

The tracker mentions we can either try to evaluate strays during scrub or we can run a thread every minute or two. As a naive approach, i felt we can do `maybe_eval_strays()` during scrubbing process. I was bit skeptical about the placement of it, after reading your comment https://github.com/ceph/ceph/pull/47649#discussion_r948731585, I would try to get it done when we scrub the inode. Makes sense?

lxbsz commented on src/mds/MDSRank.cc:
```diff
@@ -2965,7 +2965,9 @@ void MDSRank::command_scrub_start(Formatter *f,
       repair = true;
-  }
+    else if (op == "eval_stray")
+      mdcache->scan_strays();
```
Checked the scrub code and the scrub will work as:

```
ceph tell mds.<fsname>:0 scrub start <path> [scrubopts] [tag]
```

Then you can specify the `<path>` with `~mdsdir`:

```
ceph tell mds.<fsname>:0 scrub start ~mdsdir eval_stray recursive
```

And finally in `void ScrubStack::kick_off_scrubs()` or `scrub_dir_inode()` to call the `mdcache->scan_strays()` .

If the `eval_stray` option is specified the `mdcache->scan_strays()` should only eval the stray dentries, no need to check consistency for them?

@batrick @vshankar Any better idea ?

dparmar18 replied:
As discussed yesterday with @lxbsz , we can also opt for something like `ceph tell mds.<fs_name>:0 scrub eval_stray` if we don't want to interfere with scrub. (Just thinking out loud)

vshankar replied:
Cannot recursive scrub on "/" traverse ~mdsdir?

vshankar replied:
@dparmar18 ping?

mchangir replied:
> Cannot recursive scrub on "/" traverse ~mdsdir?

no ... this is not possible
`/` traverses user data and `~mdsdir` traverses system data (for lack of better word) since the strays get owned by the system and don't remain accessible to the user

batrick replied:
> Checked the scrub code and the scrub will work as:
> 
> ```
> ceph tell mds.<fsname>:0 scrub start <path> [scrubopts] [tag]
> ```
> 
> Then you can specify the `<path>` with `~mdsdir`:
> 
> ```
> ceph tell mds.<fsname>:0 scrub start ~mdsdir eval_stray recursive
> ```
> 
> And finally in `void ScrubStack::kick_off_scrubs()` or `scrub_dir_inode()` to call the `mdcache->scan_strays()` .
> 
> If the `eval_stray` option is specified the `mdcache->scan_strays()` should only eval the stray dentries, no need to check consistency for them?
> 
> @batrick @vshankar Any better idea ?

This makes sense to me except `scrub_dirfrag` should call `maybe_eval_stray` instead if it is a scrub directory dirfrag..

dparmar18 replied:
@batrick `maybe_eval_stray` would purge a dir only when that dir is marked as stray. To evaluate it we need to iterate through every item of dirfrag, mark it as stray and then call `maybe_eval_stray`. This is already being done in `MDCache::scan_stray_dir`, IMO we can just simply call it and perform evaluation.

vshankar replied:
> > Cannot recursive scrub on "/" traverse ~mdsdir?
> 
> no ... this is not possible `/` traverses user data and `~mdsdir` traverses system data (for lack of better word) since the strays get owned by the system and don't remain accessible to the user

Sure. But my point here is to additionally traverse ~mdsdir when recursive scrub is triggered on root directory. We can additionally have the user specify ~mdsdir as <path> with `scrub start`, but that's another point. BTW, has anybody tried invoke scrub on ~mdsdir from CLI to check if MDS accepts that? (seems like it does). @mchangir @dparmar18

dparmar18 replied:
@vshankar It seems MDS does accept scrub invocation from CLI (`ceph tell mds.a:0 scrub start ~mdsdir recursive`) but I'm now thinking about if it will actually evaluate strays because as @mchangir said that It might not touch system files while scrubbing.

vshankar replied:
@dparmar18 Have you tried out the changes to see what inodes srub touches once you invoke scrub start CLI with ~mdsdir?

dparmar18 replied:
scrub can iterate over ~mdsdir.

mchangir commented:
@dparmar18 I think this could be done better in `Server::handle_client_rmsnap()` or `MDSRankDispatcher::tick()`. This would avoid the requirement of the out-of-band scrub path invocation.

dparmar18 commented:
@mchangir 
> @dparmar18 I think this could be done better in `Server::handle_client_rmsnap()`

Unsure about `Server::handle_client_rmsnap`, will definitely check it out. 

>  or `MDSRankDispatcher::tick()`. This would avoid the requirement of the out-of-band scrub path invocation.

I was too thinking of running a thread periodically that would eval strays but as far as what I learnt about strays, this issue in normal scenarios should not occur that frequently. Running it recurrently might use resources unnecessarily. Makes sense?

vshankar commented:
> @mchangir
> 
> > @dparmar18 I think this could be done better in `Server::handle_client_rmsnap()`
> 
> Unsure about `Server::handle_client_rmsnap`, will definitely check it out.
> 
> > or `MDSRankDispatcher::tick()`. This would avoid the requirement of the out-of-band scrub path invocation.
> 
> I was too thinking of running a thread periodically that would eval strays but as far as what I learnt about strays, this issue in normal scenarios should not occur that frequently. Running it recurrently might use resources unnecessarily. Makes sense?

A thread dedicated for this sounds like an overkill. We do have a finisher thread and a timer task in the MDS, however, its probably _not_ wise to do blocking work via those...

dparmar18 commented:
> @mchangir
> 
> > @dparmar18 I think this could be done better in `Server::handle_client_rmsnap()`
> 
> Unsure about `Server::handle_client_rmsnap`, will definitely check it out.

Should  we do `maybe_eval_stray` on the inode we get from `mdr->client_request->get_filepath().get_ino()` every time we remove snap?

> > or `MDSRankDispatcher::tick()`. This would avoid the requirement of the out-of-band scrub path invocation.
> 
> I was too thinking of running a thread periodically that would eval strays but as far as what I learnt about strays, this issue in normal scenarios should not occur that frequently. Running it recurrently might use resources unnecessarily. Makes sense?

batrick commented:
Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.

mchangir commented:
> > @mchangir
> > > @dparmar18 I think this could be done better in `Server::handle_client_rmsnap()`
> > 
> > 
> > Unsure about `Server::handle_client_rmsnap`, will definitely check it out.
> 
> Should we do `maybe_eval_stray` on the inode we get from `mdr->client_request->get_filepath().get_ino()` every time we remove snap?

You might want to create a continuation object to eval strays, but wait until the stray is linked into the stray container dir.
Just check if there's an inode state which says that stray is created and wait on it to execute the continuation object to eval strays.

gregsfortytwo commented:
> Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.

How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?

batrick commented:
> > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> 
> How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?

No. The intended command would be:

`ceph tell mds.X scrub start ~mdsdir recursive`

That would look at the stray directories and call out (with changes in this PR)  to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.

vshankar commented:
> > > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> > 
> > 
> > How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?
> 
> No. The intended command would be:
> 
> `ceph tell mds.X scrub start ~mdsdir recursive`

Cannot we have both -- ability to scrub (and evaluate strays) when scrubbing ~mdsdir and include ~mdsdir in scrub when recursively scrubbing root?

> 
> That would look at the stray directories and call out (with changes in this PR) to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.

batrick commented:
> > > > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> > > 
> > > 
> > > How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?
> > 
> > 
> > No. The intended command would be:
> > `ceph tell mds.X scrub start ~mdsdir recursive`
> 
> Cannot we have both -- ability to scrub (and evaluate strays) when scrubbing ~mdsdir and include ~mdsdir in scrub when recursively scrubbing root?
> 
> > That would look at the stray directories and call out (with changes in this PR) to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.

We could special case root to include ~mdsdir but there would need to be a way to exclusde ~mdsdir I think.

dparmar18 commented:
> > > > > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> > > > 
> > > > 
> > > > How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?
> > > 
> > > 
> > > No. The intended command would be:
> > > `ceph tell mds.X scrub start ~mdsdir recursive`
> > 
> > 
> > Cannot we have both -- ability to scrub (and evaluate strays) when scrubbing ~mdsdir and include ~mdsdir in scrub when recursively scrubbing root?
> > > That would look at the stray directories and call out (with changes in this PR) to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.
> 
> We could special case root to include ~mdsdir but there would need to be a way to exclusde ~mdsdir I think.

I can create a flag `include_mdsdir` of type bool for that. What do you think?

vshankar commented:
> > > > > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> > > > 
> > > > 
> > > > How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?
> > > 
> > > 
> > > No. The intended command would be:
> > > `ceph tell mds.X scrub start ~mdsdir recursive`
> > 
> > 
> > Cannot we have both -- ability to scrub (and evaluate strays) when scrubbing ~mdsdir and include ~mdsdir in scrub when recursively scrubbing root?
> > > That would look at the stray directories and call out (with changes in this PR) to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.
> 
> We could special case root to include ~mdsdir but there would need to be a way to exclusde ~mdsdir I think.

Does it matter if its not excluded? (It does jack up the scrub time a bit)

batrick commented:
> > > > > > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> > > > > 
> > > > > 
> > > > > How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?
> > > > 
> > > > 
> > > > No. The intended command would be:
> > > > `ceph tell mds.X scrub start ~mdsdir recursive`
> > > 
> > > 
> > > Cannot we have both -- ability to scrub (and evaluate strays) when scrubbing ~mdsdir and include ~mdsdir in scrub when recursively scrubbing root?
> > > > That would look at the stray directories and call out (with changes in this PR) to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.
> > 
> > 
> > We could special case root to include ~mdsdir but there would need to be a way to exclusde ~mdsdir I think.
> 
> Does it matter if its not excluded? (It does jack up the scrub time a bit)

I am thinking of the potentiality that scrubbing the ~mdsdir` may cause some mds instability that needs to be avoided for some hypothetical recovery.

dparmar18 commented:
> > > > > > > Doing this work in `rmsnap` is not going to work well when the # of strays is very large. (In fact, it probably shouldn't happen in mds boot either!) There may be an argument for some incremental task that handles this periodically in the MDS but I think the correct first step here is to make scrub do this cleanup. I'm also not sure we really need an extra tag for scrub to initiate stray cleanup.
> > > > > > 
> > > > > > 
> > > > > > How do you envision scrub doing this cleanup automatically? Do we want to forcibly scan the stray directory when an admin asks to scrub a 4-file directory? That seems wildly out of proportion. A separate flag makes sense to me. Maybe we could special-case when somebody scrubs the root to also include this?
> > > > > 
> > > > > 
> > > > > No. The intended command would be:
> > > > > `ceph tell mds.X scrub start ~mdsdir recursive`
> > > > 
> > > > 
> > > > Cannot we have both -- ability to scrub (and evaluate strays) when scrubbing ~mdsdir and include ~mdsdir in scrub when recursively scrubbing root?
> > > > > That would look at the stray directories and call out (with changes in this PR) to `maybe_eval_stray`. No flag needed. It would not affect regular scrubs.
> > > 
> > > 
> > > We could special case root to include ~mdsdir but there would need to be a way to exclusde ~mdsdir I think.
> > 
> > 
> > Does it matter if its not excluded? (It does jack up the scrub time a bit)
> 
> I am thinking of the potentiality that scrubbing the ~mdsdir` may cause some mds instability that needs to be avoided for some hypothetical recovery.

I've pushed some changes, do let me know if it makes sense

lxbsz commented on src/mds/ScrubStack.cc:
```diff
@@ -281,8 +280,12 @@ bool ScrubStack::validate_inode_auth(CInode *in)
 }
 
-void ScrubStack::scrub_dir_inode(CInode *in, bool *added_children, bool *done)
+void ScrubStack::scrub_dir_inode(CInode *in, bool *added_children, bool *done, bool is_eval_stray)
```
IMO no need to add the `bool is_eval_stray` parameter, you can get this from the `header` below.

dparmar18 replied:
yeah this was something i was experimenting with, i had removed this long back. thanks.

lxbsz commented on src/mds/ScrubStack.cc:
```diff
@@ -281,8 +280,12 @@ bool ScrubStack::validate_inode_auth(CInode *in)
+  if(is_eval_stray) {
+    mdcache->maybe_eval_stray(in);
+    return;
+  }
```
You didn't call the `on_finish` after evaling the stray. This is why it stuck inifinitly.

And this also seems incorrect, IMO you should eval the stray dentries instead of the `~mdsdir` or the `stray dirs`. So you need to iterate all the stray dirs and encqueue all the stray dentries under them. And then call the  `mdcache->maybe_eval_stray(in);` for each of them ?

dparmar18 replied:
wont the recursive scrub traverse through all the stray dirs and enqueue them?

dparmar18 replied:
this is now removed

vshankar commented:
> I am thinking of the potentiality that scrubbing the ~mdsdir` may cause some mds instability that needs to be avoided for some hypothetical recovery.

Fair point. Let's run this PR through test (when its ready) and see what problem we run into when scrunning ~mdsdir.

kotreshhr commented on src/mds/ScrubStack.h:
```diff
@@ -203,15 +203,15 @@ class ScrubStack {
    * @param added_children set to true if we pushed some of our children
    * @param done set to true if we started to do final scrub
```
nit: update the comment with new argument.

dparmar18 replied:
this new arg doesn't exist anymore, closing the conversation

kotreshhr commented on src/mds/ScrubStack.h:
```diff
@@ -203,15 +203,15 @@ class ScrubStack {
    * scrub of the dirfrag.
    *
    * @param dir The dirfrag to scrub (must be auth)
    * @param done set to true if we started to do final scrub
```
nit: same as above

dparmar18 replied:
closing the conversation, this new arg has been removed

dparmar18 commented:
We can trigger stray evaluation using scrub without any additional code. By running `ceph tell mds.0 scrub start ~mdsdir recursive`, we can purge all of the strays, seems like  https://github.com/ceph/ceph/commit/83f9a49dfddcba06ab3e0cd9b860523fafad4605 does the job. I can say this after running a lot of tests (by adding big blob of data containing lots of hard links to stray dirs -> creating snap -> deleting blob of data -> deleting snap) in a loop and analyzing MDS side logs too. I will document this up.

All I feel we need is to help evaluate strays at root. So for this I just enqueued a recursive scrub on `~mdsdir` if path provided is root and used op is `eval_stray_at_root` (new op). What do you think @vshankar @lxbsz @batrick @gregsfortytwo @mchangir @kotreshhr ?

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -2958,16 +2958,22 @@ void MDSRank::command_scrub_start(Formatter *f,
   std::lock_guard l(mds_lock);
+  if (eval_stray_at_root) {
+    mdcache->enqueue_scrub("~mdsdir", tag, force, true, repair, f, on_finish);
+  }
```
I don't really see the point of this flag. The operator can just queue a scrub for `path = "~mdsdir"`.

dparmar18 replied:
You are right, that can be done but understanding your previous comments https://github.com/ceph/ceph/pull/47649#issuecomment-1235401150, https://github.com/ceph/ceph/pull/47649#issuecomment-1235862530 and after discussing with @vshankar  in a 1x1, I thought of providing this flag. Rationale being: provide way to exclude `~mdsdir` (both as convenience and safety concern). Someone might want to scrub everything(scrub + eval stray) when they mention root as path and as the scrub by default doesn't do it(i.e. include ~mdsdir) I think this flag should be a good addition.

vshankar replied:
Yeh. I recommended to add that functionality.

dparmar18 replied:
+1

batrick reviewed (dismissed):
> We can trigger stray evaluation using scrub without any additional code. By running `ceph tell mds.0 scrub start ~mdsdir recursive`, we can purge all of the strays, seems like [83f9a49](https://github.com/ceph/ceph/commit/83f9a49dfddcba06ab3e0cd9b860523fafad4605) does the job. I can say this after running a lot of tests (by adding big blob of data containing lots of hard links to stray dirs -> creating snap -> deleting blob of data -> deleting snap) in a loop and analyzing MDS side logs too. I will document this up.

Please write a test.

anthonyeleven commented on doc/cephfs/scrub.rst:
```diff
@@ -131,3 +131,15 @@ Control (ongoing) File System Scrubs
+=====================================
+
+- In order to evaluate strays i.e. purge stray directories in ``~mdsdir`` use the following command::
+
```
Is `~mdsdir` a specific CephFS thing ? With the tilde it looks like a user homedir

dparmar18 replied:
@anthonyeleven yes, `~mdsdir` contains stray dirs which is kind of CephFS exclusive

anthonyeleven replied:
Groovy, thanks.

vshankar commented on src/mds/MDSDaemon.cc:
```diff
@@ -283,7 +283,7 @@ void MDSDaemon::set_up_admin_socket()
   r = admin_socket->register_command("scrub start "
 				     "name=path,type=CephString "
-				     "name=scrubops,type=CephChoices,strings=force|recursive|repair,n=N,req=false "
+				     "name=scrubops,type=CephChoices,strings=force|recursive|repair|eval_stray_at_root,n=N,req=false "
```
Let's call this new flag `scrub_mdsdir` and document that providing this flag triggers stray reintegration.

dparmar18 replied:
done

vshankar reviewed (commented):
Mostly looks good! Good for QA round after addressing pending comments.

dparmar18 commented:
> Mostly looks good! Good for QA round after addressing pending comments.

Remember I was talking about scrub status getting unresponsive after recursive scrub on mdsdir? I actually ran a testcase to check it and found we are hitting this https://tracker.ceph.com/issues/51824. Copy pasting the [comment-9](https://tracker.ceph.com/issues/51824#note-9) here:

"I can tell this issue exists even in Quincy. The Ceph environment I used operates on`ceph 17.0.0-14883-g10f7d25d`, find below the log of the test run:
http://qa-proxy.ceph.com/teuthology/dparmar-2022-09-16_20:46:54-fs:functional-main-distro-default-smithi/7035975/teuthology.log

The testcase is aimed to evaluate strays by running recursive scrub on `~mdsdir`. It fails with the exact same stacktrace as mentioned in the description. It occurs on `scrub status` after performing the recursive scrub. Surprisingly for this testcase, `num_stray = 1847` is nothing near to the metioned value(i.e. `num_stray = 22M`). This resulting at Giga/Tera/Peta level should be anticipated. This also asserts this problem is just not pacific specific. Maybe this can also exist in octopus? I'll try running the testcase in a octopus cluster and find it out.

PS: I also found this: https://www.spinics.net/lists/ceph-users/msg71128.html on web while working on this issue. I.e. It can also lead to damaged `dir_frags`. Unsure about the severity of this but thought of sharing it."

vshankar commented:
> > Mostly looks good! Good for QA round after addressing pending comments.
> 
> Remember I was talking about scrub status getting unresponsive after recursive scrub on mdsdir? I actually ran a testcase to check it and found we are hitting this https://tracker.ceph.com/issues/51824. Copy pasting the [comment-9](https://tracker.ceph.com/issues/51824#note-9) here:
> 
> "I can tell this issue exists even in Quincy. The Ceph environment I used operates on`ceph 17.0.0-14883-g10f7d25d`, find below the log of the test run: http://qa-proxy.ceph.com/teuthology/dparmar-2022-09-16_20:46:54-fs:functional-main-distro-default-smithi/7035975/teuthology.log
> 
> The testcase is aimed to evaluate strays by running recursive scrub on `~mdsdir`. It fails with the exact same stacktrace as mentioned in the description. It occurs on `scrub status` after performing the recursive scrub. Surprisingly for this testcase, `num_stray = 1847` is nothing near to the metioned value(i.e. `num_stray = 22M`). This resulting at Giga/Tera/Peta level should be anticipated. This also asserts this problem is just not pacific specific. Maybe this can also exist in octopus? I'll try running the testcase in a octopus cluster and find it out.

This needs to be fixed before we merge and advertise this change. @dparmar18 PTAL.

> 
> PS: I also found this: https://www.spinics.net/lists/ceph-users/msg71128.html on web while working on this issue. I.e. It can also lead to damaged `dir_frags`. Unsure about the severity of this but thought of sharing it."

dparmar18 commented:
> > > Mostly looks good! Good for QA round after addressing pending comments.
> > 
> > 
> > Remember I was talking about scrub status getting unresponsive after recursive scrub on mdsdir? I actually ran a testcase to check it and found we are hitting this https://tracker.ceph.com/issues/51824. Copy pasting the [comment-9](https://tracker.ceph.com/issues/51824#note-9) here:
> > "I can tell this issue exists even in Quincy. The Ceph environment I used operates on`ceph 17.0.0-14883-g10f7d25d`, find below the log of the test run: http://qa-proxy.ceph.com/teuthology/dparmar-2022-09-16_20:46:54-fs:functional-main-distro-default-smithi/7035975/teuthology.log
> > The testcase is aimed to evaluate strays by running recursive scrub on `~mdsdir`. It fails with the exact same stacktrace as mentioned in the description. It occurs on `scrub status` after performing the recursive scrub. Surprisingly for this testcase, `num_stray = 1847` is nothing near to the metioned value(i.e. `num_stray = 22M`). This resulting at Giga/Tera/Peta level should be anticipated. This also asserts this problem is just not pacific specific. Maybe this can also exist in octopus? I'll try running the testcase in a octopus cluster and find it out.
> 
> This needs to be fixed before we merge and advertise this change. @dparmar18 PTAL.

Im on it, looking at logs and trying to figure out the root cause. Marking this PR as draft for now.
 
> > PS: I also found this: https://www.spinics.net/lists/ceph-users/msg71128.html on web while working on this issue. I.e. It can also lead to damaged `dir_frags`. Unsure about the severity of this but thought of sharing it."

dparmar18 commented:
@vshankar http://pulpito.front.sepia.ceph.com/dparmar-2022-12-08_15:20:58-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/

All tests passed, PTAL. Especially:
* second commit - Is it practical to avoids those asserts there?
* sixth commit - It's intended to fix the "IDLE (waiting for more scrubs)" issue.

dparmar18 commented:
FYI testcase in this PR tests only `ceph tell mds.X scrub start ~mdsdir recursive` and not the functionality `ceph tell mds.X scrub start / eval_stray_at_root` (`eval_stray_at_root` being the new flag to scrub root + mdsdir both at once.)

anthonyeleven commented on doc/cephfs/scrub.rst:
```diff
@@ -131,3 +131,15 @@ Control (ongoing) File System Scrubs
+
+    ceph tell mds.<fsname>:0 scrub start ~mdsdir recursive
+
+- ``~mdsdir`` is not enqueued by default when scrubbing at root. In order to perform stray evaluation
```
Suggest "at the CephFS filesystem root"

dparmar18 replied:
CepFS along with filesystem will be okay? This would be like CephFileSystem filesystem

dparmar18 replied:
im going with "at the CephFS root"

dparmar18 commented:
@vshankar As per our discussion in 1x1, here is the job link http://pulpito.front.sepia.ceph.com/dparmar-2022-12-07_18:45:43-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/7107159/

dparmar18 commented:
@vshankar 

```
2022-12-07T19:17:16.690+0000 7fc94fe00700 10 mds.0.scrubstack scrub_dir_inode done
2022-12-07T19:17:16.690+0000 7fc94fe00700 20 mds.0.scrubstack kick_off_scrubs dir inode, done
2022-12-07T19:17:16.690+0000 7fc94fe00700 20 mds.0.scrubstack dequeue [inode 0x100 [...3,head] ~mds0/ auth v38 ap=1 snaprealm=0x55c7fb42f800 f(v0 10=0+10) n(v7 rc2022-12-07T19:17:15.753753+0000 b6890 2011=2000+11)/n(v0 rc2022-12-07T19:15:41.353219+0000 11=0+11) old_inodes=1 (inest lock) (iversion lock) | dirtyscattered=0 lock=0 dirfrag=1 dirtyparent=0 dirty=1 authpin=1 scrubqueue=1 0x55c7fb34b180] from ScrubStack
2022-12-07T19:17:16.690+0000 7fc94fe00700 10 mds.0.scrubstack scrub_dirfrag [dir 0x100 ~mds0/ [2,head] auth v=13133 cv=12956/12956 dir_auth=0 ap=1+0 state=1610612737|complete f(v0 10=0+10) n(v7 rc2022-12-07T19:17:15.753753+0000 b6890 2010=2000+10) hs=10+0,ss=0+0 dirty=1 | child=1 subtree=1 dirty=1 authpin=1 scrubqueue=1 0x55c7fb326d00]
```
Is this what you were talking about? It dequeues the mdsdir inode and it also does scrub it after dequeue.

vshankar commented on src/mds/ScrubStack.cc:
```diff
@@ -62,7 +62,11 @@ void ScrubStack::dequeue(MDSCacheObject *obj)
 int ScrubStack::_enqueue(MDSCacheObject *obj, ScrubHeaderRef& header, bool top)
 {
   ceph_assert(ceph_mutex_is_locked_by_me(mdcache->mds->mds_lock));
+  bool is_inode_purging = false;
```
discussed this with @dparmar18 - my approach was to scrub an inode until it handed to the purge queue by stray manager. I guess this is fine too. We can add that later.

vshankar commented on src/mds/StrayManager.cc:
```diff
@@ -300,6 +301,11 @@ void StrayManager::enqueue(CDentry *dn, bool trunc)
   ceph_assert(in);
 
+  //remove inode from scrub stack if it is being purged
+  if(mds->scrubstack->remove_inode_if_stacked(in)) {
```
How about

```
if (obj->item_scrub.is_on_list()) {
    remove from scrub stack
}
```

dparmar18 replied:
so this would require the function `remove_inode_if_stacked` moved here, im fine with it. the reason i kept it in ScrubStack is because the function shares its relevance with ScrubStack

dparmar18 commented on src/mds/CInode.cc:
```diff
@@ -4705,7 +4705,11 @@ void CInode::validate_disk_state(CInode::validated_data *results,
       auto ino = [this]() { return in->ino(); }; // For the benefit of dout
-
+      if (in->is_mdsdir() || in->is_stray()){
+          dout(20) << "forcing backtrace as passed since mdsdir/stray dir " 
```
right, stray would have backtrace and this whole commit was a very initial experimentation. i think we can get rid of this commit

dparmar18 replied:
this code is removed as it is addressed in https://github.com/ceph/ceph/pull/48450

dparmar18 commented on src/mds/MDCache.cc:
```diff
@@ -322,7 +322,9 @@ void MDCache::remove_inode(CInode *o)
 
   // delete it
-  ceph_assert(o->get_num_ref() == 0);
+  if(!o->is_stray() && !o->is_mdsdir()) {
```
there should not be any references left when purge reaches here considering the fix provided in this PR. this is not needed anymore, will remove it.

dparmar18 replied:
removed it.

dparmar18 commented on src/mds/StrayManager.cc:
```diff
@@ -209,11 +209,11 @@ void StrayManager::_purge_stray_purged(
-	  (int)dn->is_dirty() +
-	  !!dn->state_test(CDentry::STATE_FRAGMENTING) +
-	  !!in->get_num_ref() + 1 /* PIN_PURGING */) {
+    if (!in->is_stray() && !in->is_mdsdir() && 
```
my initial impression was that stray or mdsdir, none of them should be allowed to hit the `ceph_abort_msg` cause they might be possessing some reference at the time of its purge and it was figured out that it was the scrubqueue reference, so now as it has already been addresses and fixed in this PR, this code becomes obsolete. I'll remove it.

dparmar18 replied:
removed

dparmar18 commented:
FYI: stray reintegration happens much before the purge check kicks in. So, in case we have hard links lingering in StrayManager, it would be handled without any issue. I confirmed this checking through mds logs.

dparmar18 commented:
@vshankar the second and third commit make sure that purge and scrub never run into the race condition(https://tracker.ceph.com/issues/51824) while the first commit creates a new flag to allow scrubbing mdsdir with root path

mchangir commented on qa/tasks/cephfs/mount.py:
```diff
@@ -1189,24 +1189,30 @@ def open_n_background(self, fs_path, count):
         return rproc
 
-    def create_n_files(self, fs_path, count, sync=False, dirsync=False, unlink=False, finaldirsync=False):
+    def create_n_files(self, fs_path, count, hard_links=0, sync=False,
+                       dirsync=False, unlink=False, finaldirsync=False):
```
does this change work for all other instances of `create_n_files()`  calls ... especially since a new formal parameter has been introduced for `hard_links`

dparmar18 replied:
I'll check all the instances, thanks for pointing this out

dparmar18 replied:
@mchangir I checked all instances and I found `fs_path` and `count` being used in most cases as positional args while other parameters being passed as keyword args. Since there is no change in order of `fs_path` and `count`, all the instances look safe however there is an exception `self.mount_a.create_n_files("testdir/file2", 5, True)` in test_client_limits.py where the third arg which is `True` would mean `hard_links=True`, for this either I can make change to that particular line or I can shift this new arg(`hard_links`) to the end of the param list. What do you think?

mchangir replied:
as a matter of good practice, anything new should be appended at the end as far as possible

dparmar18 replied:
got it, thanks! @mchangir

dparmar18 replied:
done

vshankar commented on src/mds/MDSRank.cc:
```diff
@@ -2970,16 +2970,22 @@ void MDSRank::command_scrub_start(Formatter *f,
+  if (scrub_mdsdir) {
+    mdcache->enqueue_scrub("~mdsdir", tag, false, true, false, f, on_finish);
+  }
   mdcache->enqueue_scrub(path, tag, force, recursive, repair, f, on_finish);
```
As you noted, this could segfault if `on_finish` completion callback is invoked before this is executed. You can use `MDSGather` and create new context for each `enqueue_scrub` invocation and then `activate`.

dparmar18 replied:
this change has been incorporated. thanks

vshankar commented on src/mds/ScrubStack.cc:
```diff
@@ -663,7 +663,11 @@ void ScrubStack::scrub_status(Formatter *f) {
 
-    f->dump_string("tag", header->get_tag());
+    if (mdcache->get_inode(header->get_origin())->is_mdsdir() && tag.empty()) {
+      f->dump_string("tag", "scrub_mdsdir_at_root");
```
This is going to confuse users _a lot_. If there any need to show that mdsdir is included in scrub. Maybe introducing a `flags` section in the json and including it there?

dparmar18 replied:
Sounds good, will try that out

dparmar18 replied:
this change has been incorporated. thanks

dparmar18 replied:
we already have `options`, made changes in https://github.com/ceph/ceph/pull/47649/commits/3badfe7171cad89e25255f805ec241c3d12166f3 to print `scrub_mdsdir`

dparmar18 commented:
resolved 

```
tasks/cephfs/test_scrub_checks.py:301: error: "None" has no attribute "set_allow_new_snaps"
tasks/cephfs/test_scrub_checks.py:304: error: "None" has no attribute "run_shell"
tasks/cephfs/test_scrub_checks.py:305: error: "None" has no attribute "mountpoint"
tasks/cephfs/test_scrub_checks.py:306: error: "None" has no attribute "create_n_files"
tasks/cephfs/test_scrub_checks.py:308: error: "None" has no attribute "run_shell"
tasks/cephfs/test_scrub_checks.py:310: error: "None" has no attribute "run_shell"
tasks/cephfs/test_scrub_checks.py:311: error: "None" has no attribute "run_shell"
tasks/cephfs/test_scrub_checks.py:313: error: "None" has no attribute "rank_tell"
tasks/cephfs/test_scrub_checks.py:320: error: "None" has no attribute "run_scrub"
tasks/cephfs/test_scrub_checks.py:325: error: "None" has no attribute "wait_until_scrub_complete"
tasks/cephfs/test_scrub_checks.py:327: error: "None" has no attribute "rank_tell"
```

faced in make check

dparmar18 commented:
last push resolved conflict

dparmar18 commented:
jenkins test api

dparmar18 commented:
jenkins test make check arm64

dparmar18 commented:
last push:
- Addressed https://github.com/ceph/ceph/pull/47649#discussion_r1072003246
- Fixed doc (had removed some text unknowingly)
- Improved last commit message

dparmar18 commented:
@vshankar ran this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-02-22_06:04:20-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/

problem `teuthology.exceptions.CommandCrashedError: Command crashed: 'ceph-dencoder type inode_backtrace_t import - decode dump_json'` still exists

dparmar18 commented:
> @vshankar ran this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-02-22_06:04:20-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/
> 
> problem `teuthology.exceptions.CommandCrashedError: Command crashed: 'ceph-dencoder type inode_backtrace_t import - decode dump_json'` still exists

Seems like I need to rebase and test again as this PR https://github.com/ceph/ceph/pull/49842 got merged three weeks ago and my last rebase was just an hour before it got merged :)

vshankar commented:
> > @vshankar ran this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-02-22_06:04:20-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/
> > problem `teuthology.exceptions.CommandCrashedError: Command crashed: 'ceph-dencoder type inode_backtrace_t import - decode dump_json'` still exists
> 
> Seems like I need to rebase and test again as this PR #49842 got merged three weeks ago and my last rebase was just an hour before it got merged :)

ah! please do the needful.

dparmar18 commented:
> > > @vshankar ran this PR http://pulpito.front.sepia.ceph.com/dparmar-2023-02-22_06:04:20-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/
> > > problem `teuthology.exceptions.CommandCrashedError: Command crashed: 'ceph-dencoder type inode_backtrace_t import - decode dump_json'` still exists
> > 
> > 
> > Seems like I need to rebase and test again as this PR #49842 got merged three weeks ago and my last rebase was just an hour before it got merged :)
> 
> ah! please do the needful.

Rebased

vshankar commented:
@dparmar18 I have included this is my testing branch. Could you ensure the jenkins tests pass (or at least ensure they aren't related).

dparmar18 commented:
> @dparmar18 I have included this is my testing branch. Could you ensure the jenkins tests pass (or at least ensure they aren't related).

current failures dont have logs, let me rebase and run tests again

dparmar18 commented:
rebased

dparmar18 commented:
jenkins test api

dparmar18 commented:
https://jenkins.ceph.com/job/ceph-api/51838/consoleText

dparmar18 commented:
API test fails at `[errno 2] RADOS object not found (error connecting to the cluster)`, facing same issue in other PRs too. Although to be completely sure, i had scheduled a run and it's all green http://pulpito.front.sepia.ceph.com/dparmar-2023-03-03_10:04:35-fs:functional-wip-dparmar-MDS-53724-distro-default-smithi/

dparmar18 commented:
jenkins test api

dparmar18 commented:
> @dparmar18 I have included this is my testing branch. Could you ensure the jenkins tests pass (or at least ensure they aren't related).

@vshankar passed

vshankar commented:
> https://pulpito.ceph.com/vshankar-2023-03-16_06:43:29-fs-wip-vshankar-testing-20230315.085747-testing-default-smithi/

This change is good to merge once I update the run wiki with the known failures.

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Main#30-March-2023
